CDCgov/prime-reportstream
# 5839 - Generate OBR segments from FHIR

This PR ...
Adds the missing mappers from the original HL7 -> FHIR so that we can then map them back to the regenerated HL7.
It also factors out coded element into a separated helper. Detailed list of changes bellow.


Test Steps:
Run this command to generate the HL7`./gradlew primeCLI --args='fhirdata --input-file "src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir" -s metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml'`
You will notice these differences:
- OBR.3.4 doesn't include ISO at the end because it doesn't actually have a universal id. This is purposeful
- OBR.4 doesn't include the non 2.5.1 HL7 at the end
- OBR.16 Ordering Provider has MD in a different spot
- OBR.26 doesn't include the non 2.5.1 HL7 at the end
- OBR.29.2.2 is missing "RMI" (I can't get this to map from the original HL7 to FHIR)
- OBR.29.2.4 doesn't include ISO at the end because it doesn't actually have a universal id. This is purposeful
- OBR.31 doesn't include the non 2.5.1 HL7 at the end


## Changes
- Adds mappers from HL7 to FHIR for OBR 19, 20, 21, 26, and 29
- Generates the HL7 OBR segment from the FHIR DiagnosticReport and ServiceRequest
- Adds a coded element datatype helper

## Checklist

### Testing

### Process

## Linked Issues
- 5839

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
